### ISSUE TITLE
PDE no longer passing –external

### ISSUE THREAD

**Issue by jeremydouglass** [processing/processing#5843]
## Description
I have sketches that I've been using for several years that can be run either from PDE (manually) or from processing-java (automatically, with optional arguments). I use PApplet "external" flag ([documented here](http://processing.github.io/processing-javadocs/core/processing/core/PApplet.html#main-java.lang.String:A-))

> Parameters used by Processing when running via the PDE
>  --external             set when the applet is being used by the PDE

...to detect whether the sketch was launched by PDE or not, as in this sketch:

```
boolean PDE;
void setup() {
  String cmd = System.getProperty("sun.java.command");
  println(cmd);
  if (cmd.contains("--external")) {
    PDE=true;
  } else {
    PDE=false;
  }
  println(PDE);
}
```

That works up to 3.4 – when run from PDE the output contains “–external” as per the documentation, and the flag is set.

> processing.core.PApplet --location=1213,185 --external --display=1 --sketch-path=/Users/student/Downloads/Test Test

Now, in 3.5.3, that doesn't work. When run from PDE, the “external” flag is missing.

> processing.core.PApplet --display=1 --sketch-path=/Users/student/Downloads/Test Test

Any idea why / how this might have happened? Is it a bug / regression, or a change in Processing, or the bundled Java? I’m not even sure where to start looking.

## Expected Behavior
PDE passes flag, giving evidence that it is the runtime context.

## Current Behavior
In PDE 3.5.3, the external flag is no longer observed in `System.getProperty("sun.java.command")`

## Steps to Reproduce
1. Run the test sketch in PDE 3.4 -- external flag is observed
2. Run the test sketch in PDE 3.5.3 -- external flag is not observed

## Your Environment
* Processing version: 3.4 / 3.5.3
* Operating System and OS version: MacOS 10.12.6, also tested on MacOS 13 and Windows 10

-  Old discussion: https://forum.processing.org/two/discussion/23999/detect-command-line-invocation-of-a-sketch
-  Current forum thread: https://discourse.processing.org/t/cant-detect-pde-vs-command-line-pde-no-longer-passing-external/10508

**Comment by jeremydouglass** [comment]
Any suggestions?

I'm having my students use 3.4 instead of 3.5.3 in the interim, but this is disrupting assignments for about 80 students, so I'm hoping to at least deploy a workaround for 3.5.3 if possible. I'm open to doing more research if there are any suggestions on where to start looking in the commits -- blindly bisecting seems like it could take a long time.

**Comment by jeremydouglass** [comment]
Any suggestions on this? Even just which area of the code to start looking in?

My best guess is that it is some commit in Runner.java from the past two years, but I'n having a hard time locating it and honestly don't understand this part of the codebase very well.

https://github.com/processing/processing/commits/349f413a3fb63a75e0b096097a5b0ba7f5565198/java/src/processing/mode/java/runner/Runner.java

**Comment by benfry** [comment]
Hm, `Runner.java` is a good guess, though it doesn't look like it's changed between 3.4 and 3.5.3. 

My guess at the moment would be the `RunnerListenerEdtAdapter` addition may have broken something since it changed how `Editor` objects were passed around. But that's just a guess.

**Comment by benfry** [comment]
Found and fixed for 4.0 alpha 4: https://github.com/processing/processing4/issues/158

(Turns out the guess was correct; the adapter was wrapping the JavaEditor object, so it wasn't being recognized as *running* from an Editor window. https://github.com/processing/processing4/commit/fe5f7d568da3fd9e513876f117a33f97eb146d8a)
